### PR TITLE
Driver for Anello EVK

### PR DESCRIPTION
**Pull request by gssammy**
# Description
Implements a "driver" in the ArduPilot code base that can be used to communicate with an AnelloEVK over UART. 

# General Implementation Details:
- Uses the AP built-in Serial Manager to receive UART packets
- store packets in the buffer
- look for the start of a packet
- read the rest of the bits based on expected length based on the type of msg (might change this to be based on length data, which straddles byte)
- store in appropriate struct
- parse struct data to pass to AP as sensor readings

# Important Files:
- [AP_ExternalAHRS_AnelloEVK.cpp](https://github.com/skywaysinc/ardupilot-anello-integration/pull/1/files#diff-dc8f552a61afc2ae6adce48d1368dfe7cf0d7b2484e4ac02d4c3d65b23158f2e)
- [AP_ExternalAHRS_AnelloEVK.h](https://github.com/skywaysinc/ardupilot-anello-integration/pull/1/files#diff-c0190e219b4f8d08efb42bbb8550477c609745dd825d4a7786c939995b38440d)

# TODO
- [ ] Deal with GPS message
- [ ] Deal with INS message
- [ ] Get feedback from AP persons
- [ ] Tabletop test communication and parsing with EVK

### CODE REVIEW

File: Tools/environment_install/install-prereqs-ubuntu.sh [hunk 1]
```diff
@@ -252,9 +252,9 @@ $APT_GET install $BASE_PKGS $SITL_PKGS $PX4_PKGS $ARM_LINUX_PKGS $COVERAGE_PKGS
 # Update Pip and Setuptools on old distro
 if [ ${RELEASE_CODENAME} == 'xenial' ] || [ ${RELEASE_CODENAME} == 'disco' ] || [ ${RELEASE_CODENAME} == 'eoan' ]; then
     # use fixed version for package that drop python2 support
-    $PIP install --user -U pip==20.3 setuptools==44.0.0
+    $PIP install -U pip==20.3 setuptools==44.0.0
 fi
-$PIP install --user -U $PYTHON_PKGS
```
Comment (magate) [inline Tools/environment_install/install-prereqs-ubuntu.sh:257]: Why this change?

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS.h [hunk 2]
```diff
@@ -13,7 +13,7 @@
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 /*
-  suppport for serial connected AHRS systems
```
Comment (magate) [inline libraries/AP_ExternalAHRS/AP_ExternalAHRS.h:16]: nit: I think spelling and white space are supposed to be separate commits from code changes.

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.cpp [hunk 3]
```diff
@@ -0,0 +1,571 @@
+// Skyways
```
Comment (magate) [inline libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.cpp:1]: nit: Are you going to push this upstream? I'm not sure what we do about the `// Skyways` in that case.

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.cpp [hunk 4]
```diff
@@ -0,0 +1,571 @@
...
+            }
+            break;
+        case ParseState::WaitingFor_PayloadLength:
+            message_in.packet.length[message_in.index++] = b;
+            if (message_in.index == 2) {
+                message_in.packet.length[2] = message_in.packet.length[0] << 8 | message_in.packet.length[1] ;
+                message_in.index = 0;
+                message_in.state = ParseState::Reading_Data;
+            }
+            //TODO put in check so we don't get stuck here.
```
Comment (magate) [inline libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.cpp]: What is with the `//TODO`?

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.cpp [hunk 5]
```diff
@@ -0,0 +1,571 @@
...
+    }
+    if (!filterStatus.flags.initalized) {
+        flags |= EKF_UNINITIALIZED;
+    }
+
+    // send message
+    const float vel_gate = 4; // represents hz value data is posted at
+    const float pos_gate = 4; // represents hz value data is posted at
+    const float hgt_gate = 4; // represents hz value data is posted at
+    const float mag_var = 0; //we may need to change this to be like the other gates, set to 0 because mag is ignored by the ins filter in vectornav
```
Comment (magate) [inline libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.cpp]: Is this comment accurate?

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.h [hunk 6]
```diff
@@ -74,15 +74,20 @@ class AP_ExternalAHRS_AnelloEVK: public AP_ExternalAHRS_backend
      * https://docs-a1.readthedocs.io/en/latest/communication_messaging.html#binary-data-output-messages
      */
     struct AnelloEVK_Packet {
-        uint16_t length[3];
-        uint8_t payload[64];
-        uint8_t CRC[3];
+        uint8_t preamble;               ///< 0xD3 indicating start of Anello Packet
```
Comment (magate) [inline libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.h]: What is with this comment style?

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.h [hunk 7]
```diff
@@ -148,6 +153,7 @@ class AP_ExternalAHRS_AnelloEVK: public AP_ExternalAHRS_backend
     double extract_double(const uint8_t* data, uint8_t offset) const;
     float extract_float(const uint8_t* data, uint8_t offset) const;
     Quaternion populate_quaternion(const uint8_t* data, uint8_t offset) const;
+    uint8_t data_packet_length(const AnelloEVK_Packet & packet) const;
```
Comment (magate) [inline libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.h]: Is this a bitwise and? Is it supposed to be passed by reference? See 571 of the cpp file.

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.cpp [hunk 8]
```diff
@@ -567,5 +568,10 @@ double AP_ExternalAHRS_AnelloEVK::extract_double(const uint8_t *data, uint8_t of
     return *reinterpret_cast<double*>(&tmp);
 }
 
+uint8_t AP_ExternalAHRS_AnelloEVK::data_packet_length(const AnelloEVK_Packet &packet) const {
```
Comment (magate) [inline libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.cpp]: `const AnelloEVK_Packet &packet` is not the same as in the header file.

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.cpp [hunk 9]
```diff
@@ -20,411 +20,186 @@ Driver for reading UART data from connected Anello EVK
...
-enum class INSPacketField {
-    ACCEL = 0x04,
-    GYRO = 0x05,
-    QUAT = 0x0A,
-    MAG = 0x06,
-    PRESSURE = 0x17
+/**
+ * @brief Header struct used for all Anello packets
+ *
+ * see:https://docs-a1.readthedocs.io/en/latest/communication_messaging.html#binary-data-output-messages
```
Comment (magate) [inline libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.cpp:38]: nit: you change this format in the next comment

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.cpp [hunk 10]
```diff
@@ -20,411 +20,186 @@ Driver for reading UART data from connected Anello EVK
...
-    baudrate = sm.find_baudrate(AP_SerialManager::SerialProtocol_AHRS, 0);
-    port_num = sm.find_portnum(AP_SerialManager::SerialProtocol_AHRS, 0);
-
     if (!uart) {
         GCS_SEND_TEXT(MAV_SEVERITY_INFO, "ExternalAHRS no UART");
         return;
     }
 
+    baudrate = sm.find_baudrate(AP_SerialManager::SerialProtocol_AHRS, 0);
+    port_num = sm.find_portnum(AP_SerialManager::SerialProtocol_AHRS, 0);
```
Comment (magate) [inline libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.cpp:79]: Why did these move under the `(!uart)` check?

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.cpp [hunk 11]
```diff
@@ -20,411 +20,186 @@ Driver for reading UART data from connected Anello EVK
...
-            imu_data.gyro = populate_vector3f(packet.payload, i+2);
-            break;
-        }
-        // Quaternion
-        case INSPacketField::QUAT: {
-            imu_data.quat = populate_quaternion(packet.payload, i+2);
-            break;
-        }
+        // Process packet if they are equal
+        if (crc == supplied_crc ) {
```
Comment (magate) [inline libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.cpp:135]: nit: format `(crc == supplied_crc)`

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.cpp [hunk 12]
```diff
@@ -436,12 +211,12 @@ int8_t AP_ExternalAHRS_AnelloEVK::get_port(void) const
 bool AP_ExternalAHRS_AnelloEVK::healthy(void) const
 {
     uint32_t now = AP_HAL::millis();
-    return (now - last_ins_pkt < 40 && now - last_gps_pkt < 500 && now - last_filter_pkt < 500);
+    return (now - last_imu_pkt_ms < 40);
```
Comment (magate) [inline libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.cpp:214]: Why did you get rid of the 500 ms packet check?

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.cpp [hunk 13]
```diff
@@ -525,51 +284,13 @@ void AP_ExternalAHRS_AnelloEVK::send_status_report(mavlink_channel_t chan) const
...
-    return *reinterpret_cast<float*>(&tmp);
-}
-
-double AP_ExternalAHRS_AnelloEVK::extract_double(const uint8_t *data, uint8_t offset) const
-{
-    uint64_t tmp = be64toh_ptr(&data[offset]);
-
-    return *reinterpret_cast<double*>(&tmp);
-}
+    // const struct AnelloPacket_IMU_pkt &pkt = *(struct AnelloPacket_IMU_pkt *)last_imu_pkt;
```
Comment (magate) [inline libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.cpp:287]: nit: what are all these commented out lines?

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.h [hunk 14]
```diff
@@ -30,6 +30,8 @@ Driver for reading UART data from connected Anello EVK
 /**
  * @brief AnelloEVK implementation using the already existing External AHRS
  * pipeline
+ *
+ * This implementation follows the implementation for the VectorNav driver.
```
Comment (magate) [inline libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.h:34]: nit: Put this in the initial commit.

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.h [hunk 15]
```diff
@@ -49,120 +51,26 @@ class AP_ExternalAHRS_AnelloEVK: public AP_ExternalAHRS_backend
...
     AP_HAL::UARTDriver *uart;        ///< Pointer to the uart accessor of the serial manager
     bool port_open = false;          ///< Flag to know if to read data from port
-    const uint8_t GPS_MSG = 0xA2;    ///< Byte #2's value of the data section of the packet indicating GPS data
-    const uint8_t IMU_MSG = 0xA1;    ///< Byte #2's value of the data section of the packet indicating IMU data
-    const uint8_t INS_MSG = 0xA3;    ///< Byte #2's value of the data section of the packet indicating INS data
-    const uint8_t PREAMBLE = 0xD3;   ///< Value which Anello EVK sends before sending packet
     HAL_Semaphore sem;               ///< ???
     int8_t port_num;                 ///< which port the device is connected too
+    struct IMU_pkt *last_imu_pkt;
+    uint16_t buf_size = 70;          ///< largest anello packet is 70 bytes
```
Comment (magate) [inline libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.h:64]: Does this need a link to a reference?

**Review by magate — COMMENTED** [review]
General nits:

* These commits are huge. I can't follow the logic behind each one.
* Commit messages need work.
* Is there any way to write a test for this code or is it all going to be hardware tested?

I don't understand the majority of this. I'll need to give it a second review, potentially a third.

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS.h (hunk 2 above)
Comment (hendjoshsr71) [reply libraries/AP_ExternalAHRS/AP_ExternalAHRS.h:16]: yeah usually, but nobody will get too angry if throw in a couple in "nearby" lines

But please dont go fixing all the whitespace and spacing mistakes.

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS.h [hunk 16]
```diff
@@ -48,6 +48,7 @@ class AP_ExternalAHRS {
         None   = 0,
         VecNav = 1,
         LORD = 2,
+        AnelloEVK = 3, // Skyways
```
Comment (hendjoshsr71) [inline libraries/AP_ExternalAHRS/AP_ExternalAHRS.h:51]:
```suggestion
        AnelloEVK = 3
```

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.cpp (hunk 3 above)
Comment (hendjoshsr71) [reply libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.cpp:1]: You are allowed to add a copyright Skyways if you would like attribution to the code added by Skyways. 

The license will still be GPLV3 

There are examples of Intel doing  this in the code if you search for "Intel".

File: libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.h [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+    // check for new data
+    void update() override {
+        check_uart();
+    };
+
+private:
+
+    AP_HAL::UARTDriver *uart;        ///< Pointer to the uart accessor of the serial manager
+    bool port_open = false;          ///< Flag to know if to read data from port
+    HAL_Semaphore sem;               ///< ???
```
Comment (hendjoshsr71) [inline libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloEVK.h:61]: the semaphore is the object to thread lock pieces of code when it is in context

**Review by hendjoshsr71 — COMMENTED** [review]
i just followed up on some of mark's comments. 

I didn't look over the code much since it is likely older then what you are testing? 

Any updates on it?

**Comment by gssammy** [comment]
This is no longer the primary method of implementation.
